nhrpd/nhrp_interface.c

```diff
@@ -136,30 +179,50 @@ static void nhrp_interface_interface_notifier(struct notifier_block *n,
 	}
 }
 
-static void nhrp_interface_update_nbma(struct interface *ifp)
+void nhrp_interface_update_nbma(struct interface *ifp,
+				struct nhrp_gre_info *gre_info)
 {
 	struct nhrp_interface *nifp = ifp->info, *nbmanifp = NULL;
 	struct interface *nbmaifp = NULL;
 	union sockunion nbma;
+	int ret;
 
 	sockunion_family(&nbma) = AF_UNSPEC;
 
 	if (nifp->source)
-		nbmaifp = if_lookup_by_name(nifp->source, VRF_DEFAULT);
+		nbmaifp = if_lookup_by_name(nifp->source, nifp->link_vrf_id);
 
 	switch (ifp->ll_type) {
 	case ZEBRA_LLT_IPGRE: {
 		struct in_addr saddr = {0};
-		netlink_gre_get_info(ifp->ifindex, &nifp->grekey,
-				     &nifp->linkidx, &saddr);
-		debugf(NHRP_DEBUG_IF, "%s: GRE: %x %x %x", ifp->name,
-		       nifp->grekey, nifp->linkidx, saddr.s_addr);
-		if (saddr.s_addr != INADDR_ANY)
-			sockunion_set(&nbma, AF_INET, (uint8_t *)&saddr.s_addr,
-				      sizeof(saddr.s_addr));
-		else if (!nbmaifp && nifp->linkidx != IFINDEX_INTERNAL)
-			nbmaifp =
-				if_lookup_by_index(nifp->linkidx, VRF_DEFAULT);
+
+		if (gre_info) {
+			nifp->i_grekey = gre_info->ikey;
+			nifp->o_grekey = gre_info->okey;
+			nifp->link_idx = gre_info->ifindex_link;
+			nifp->link_vrf_id = gre_info->vrfid_link;
+			saddr.s_addr = gre_info->vtep_ip.s_addr;
+			ret = 0;
+		} else {
+			ret = nhrp_send_zebra_gre_request(ifp,
```

Comment:
this is a little hard to understand - it _sounds_ like it is sending a request to zebra, and it is, but it _also_ looks some info up locally, in nhrpd, and returns it directly. that sort of feels like a side-effect? why not just do the lookup here, where you want to see the information, and simplify the "send_zebra_gre" api so that it just ... sends to zebra?

Reply:
you re right. I simplified it.